### PR TITLE
Context service

### PR DESCRIPTION
**Pull request by fedefernandez**
This PR makes the `Span[F]` context abstract so that you can make your service dependant on any "context."

The idea comes from #1328

Pending tasks:
* [x] Add tests for a custom context
* [x] Update docs -> Will do in the following PR

Currently, we provide some methods in the companion object for getting a parameterized algebra over `Span[F]` with `Kleisli`

Given:

```scala
trait MyService[F]
```

The method `MyService.tracingClient[F[_]]` provides a:

```scala
Resource[F, MyService[Kleisli[F, Span[F], *]]]
```

And the method `MyService.bindTracingService[F[_]]` asks for an implicit instance implementation of the algebra parametized with `Kleisli[F, Span[F], *]` (`MyService[Kleisli[F, Span[F], *]]`).

In this PR, we're adding alternative methods to those "tracing" ones where you can specify any type as a context as soon you provide a couple of implicit implementations.

The new method `MyService.contextClient[F[_], C]` will require an implicit `ClientContext[F, C]`:

https://github.com/higherkindness/mu-scala/blob/9a0e04b85d7e3726596256b504daa1dfd82f3ee9/modules/service/src/main/scala/higherkindness/mu/rpc/internal/context/ClientContext.scala#L6-L15

While the new method `MyService.bindContextService[F[_]]` will require an implicit `ServerContext[F, C]`

https://github.com/higherkindness/mu-scala/blob/9a0e04b85d7e3726596256b504daa1dfd82f3ee9/modules/service/src/main/scala/higherkindness/mu/rpc/internal/context/ServerContext.scala#L6-L10

I've chosen `trait`s because the instances need to be parameterized over `F` and `C` but allow any `Req` and `Res` (this is passed in the `apply`). I'm open to suggestions for a more ergonomic or better way to provide these implementations.

### CODE REVIEW

**Review by AdrianRaFo — DISMISSED** [review]

**Comment by cb372** [comment]
LGTM. It's probably worth updating the docs (e.g. https://higherkindness.io/mu-scala/guides/distributed-tracing) and the [examples repo](https://github.com/higherkindness/mu-scala-examples/tree/master/tracing) to reflect these changes. And maybe add a small how-to guide to the docs, showing how to use this feature to access the metadata?

**Comment by fedefernandez** [comment]
@cb372 yes! Initially, I was thinking about adding that in this PR, but this is getting bigger, and I prefer to keep this for code/implementation discussions and then open an issue to address that. Good point with the guide

File: modules/fs2/src/main/scala/higherkindness/mu/rpc/internal/client/fs2/calls.scala [hunk 1]
```diff
@@ -86,72 +85,64 @@ object calls {
...
-            input.translate(Kleisli.applyK[F, Span[F]](span))
-          clientStreaming[F, Req, Res](
-            streamF,
-            descriptor,
-            channel,
-            options,
-            headers
-          )
-        }
+  )(implicit C: ClientContext[F, C]): Kleisli[F, C, Res] =
```
Comment (juanpedromoreno) [inline modules/fs2/src/main/scala/higherkindness/mu/rpc/internal/client/fs2/calls.scala:93]: _question_: would it be better if we rename this `C `implicit `val` differently from the `C[F]` implicit bound? (one or the other)

(same suggestion for the rest of the methods too)

**Review by juanpedromoreno — COMMENTED** [review]
Besides my minor comment, shall we tag this PR as `breaking-change`?

**Comment by fedefernandez** [comment]
@juanpedromoreno this should be backward compatible (besides the deprecated warning) so I think this should be fine without the `breaking-change` tag. I'll take a look at the val names.

**Review by juanpedromoreno — APPROVED** [review]

**Pull request merged**